libs/contrib/Data/String/Parser.idr

```diff
@@ -290,3 +315,48 @@ integer : Monad m => ParseT m Integer
 integer = do minus <- succeeds (char '-')
              x <- some digit
              pure $ if minus then (intFromDigits x)*(-1) else intFromDigits x
+
+
+||| Parse repeated instances of at least one `p`, separated by `s`,
+||| returning a list of successes.
+|||
+||| @ p the parser for items
+||| @ s the parser for separators
+export
+covering
+sepBy1 : Monad m => (p : ParseT m a)
+                 -> (s : ParseT m b)
+                 -> ParseT m (List1 a)
+sepBy1 p s = [| p ::: many (s *> p) |]
+
+||| Parse zero or more `p`s, separated by `s`s, returning a list of
+||| successes.
+|||
+||| @ p the parser for items
+||| @ s the parser for separators
+export
+covering
+sepBy : Monad m => (p : ParseT m a)
+                -> (s : ParseT m b)
+                -> ParseT m (List a)
+sepBy p s = (do a <- (p `sepBy1` s)
+                pure $ forget a) <|> pure []
```

Comment:
You can also implement it with ```optionMap [] forget (p `sepBy1` s)```

Reply:
Come to think of it, all instances of `bar <|> pure foo` here can be rewritten as `optional foo bar` too, not sure if it's beneficial performance-wise though.

Reply:
Thinking about it a bit, does this combinator make sense in general? I.e., do you really want to get `[]` out of parsing `"abc"` with ```integer `sepBy` char ','```?

Reply:
Yeah I found that a little weird too, but I just copied it from [Lightyear](https://github.com/ziman/lightyear/blob/f737e25a09c1fe7c5fff063c53bd7458be232cc8/Lightyear/Combinators.idr#L77) and it seems [parsec](http://hackage.haskell.org/package/parsec-3.0.0/docs/Text-Parsec-Combinator.html#v:sepBy) behaves the same.

Reply:
> do you really want to get [] out of parsing "abc" with integer `sepBy` char ','?

Yeah, why not? If you want to insist on getting at least one value then you should use `sepBy1`.

Reply:
updated to use:
```
optionMap [] forget (p `sepBy1` s)
```
because that looks much better.

As regards what this parser _should_ be, good question. It seems like the ideal behaviour would be something like:
```
sepBy p s = (do pure $ forget !(p `sepBy1` s)) <|> (spaces *> pure [])
```
so there's a difference between the `sepBy1` parsing failing in that case that the text is empty vs the case that the text is garbage. There may be a reason the other implementations didn't do it this way that I'm missing, eg. I can imagine that `spaces` might be a bad assumption.

Reply:
I would expect

```idris
id <$ char '[' <*> sepBy p (char ',') <* char ']'
```

to successfully parse lists of `p` values, including the empty list "[]".

Reply:
Also it's probably a bad idea to arbitrarily munch some of the space:
someone may actually care very much about consuming space themselves
(e.g. because their syntax is indentation-based)

Reply:
@gallais so the example you gave surrounded by `[]` seems to behave as you expect (I slightly changed it below, hopefully not in important ways), it's the one that @clayrat suggested that's a bit strange:

```
Main> parse (char '[' *> sepBy letter (char ',') <* char ']') "[a,b]"
Right (['a', 'b'], 5)  -- ✅
Main> parse (char '[' *> sepBy letter (char ',') <* char ']') "[a]"
Right (['a'], 3) -- ✅
Main> parse (char '[' *> sepBy letter (char ',') <* char ']') "[]"
Right ([], 2) -- ✅
Main> parse (char '[' *> sepBy letter (char ',') <* char ']') "[1]"
Left (fastConcat ["Parse failed at position ", "1", ": ", "char ']'"]) -- ✅
Main> parse (sepBy letter (char ',')) "a"
Right (['a'], 1) -- ✅
Main> parse (sepBy letter (char ',')) "a,b"
Right (['a', 'b'], 3) -- ✅
Main> parse (sepBy letter (char ',')) "1"
Right ([], 0) -- 🤨
Main> parse (sepBy letter (char ',')) "]]]"
Right ([], 0) -- 🤨
```

I would have expected the last 2 to be parse errors and not empty lists. It seems lightyear and parsec behave this way, so I'm ok with it, it's just a little surprising to me.

Reply:
Something like `parse (commaSep integer) "[a,b,c]"` might be even more puzzling potentially

Reply:
The pointer into the string returned at the end is still at `0` which
indicates nothing has been consumed. If you meant for something to only
succeed if it parses a whole input then you should have a parser of the
form `p <* eof`!

Reply:
Oh! I hadn't spotted that `0`, makes much more sense now 😄